projects/swimlane/ngx-ui/src/lib/components/calendar/calendar.component.ts

```diff
@@ -122,31 +210,138 @@ export class CalendarComponent implements OnInit, ControlValueAccessor {
   _value: Date;
   weeks: Month;
 
+  currentView: View;
+  monthsList: string[] = moment.monthsShort();
+  // today's date
+  current: moment.Moment;
+  startYear: number;
+  _minView: View;
+  _defaultView: View;
+
+  get minView(): View {
+    // default: 'date'
+    return this._minView ? this._minView : 'date';
+  }
+
+  @Input('minView')
+  set minView(val: View) {
+    this._minView = val;
+    this.validateView();
+  }
+
+  get defaultView(): View {
+    return this._defaultView ? this._defaultView : this.minView;
+  }
+
+  @Input('defaultView')
+  set defaultView(val: View) {
+    this._defaultView = val;
+    this.validateView();
+  }
+  
+  changeViews(): void {
+    if (this.currentView === 'date') {
+      this.currentView = 'month';
+    } else if (this.currentView === 'month') {
+      this.currentView = 'year';
+    } else if (this.currentView === 'year') {
+      this.currentView = this.minView;
+    }
+    this.weeks = getMonth(this.activeDate);
+  }
+
+  validateView(): void {
+    const viewsList = ['date', 'month', 'year'];
+    // date time picker precision validation
+    if (!viewsList.includes(this.minView)) {
+      this.minView = 'date';
+    }
+    // defaultView cannot be below minView
+    if (viewsList.indexOf(this.minView) > viewsList.indexOf(this.defaultView)) {
+      this.defaultView = this.minView;
+    }
+    this.currentView = this.defaultView;
+  }
+
   ngOnInit(): void {
     this.activeDate = this.createMoment(this.value);
     this.weeks = getMonth(this.activeDate);
+    this.monthsList = moment.monthsShort();
+    this.current = this.activeDate;
+    this.startYear = this.current.year() - (this.current.year() % 20) + 1;
+    this.validateView();
   }
 
   getDayActive(date: moment.Moment): boolean {
     return date.isSame(this.value, 'day');
   }
 
-  getDayDisabled(date: moment.Moment): boolean {
-    if (this.disabled) return true;
-    if (!date) return false;
+  getMonthActive(month: string): boolean {
+    const date = moment(this.value).clone().month(month);
+    return date.isSame(this.value, 'month') && date.isSame(this.activeDate, 'year');
+  };
 
-    const isBeforeMin = this.minDate && date.isBefore(this.parseDate(this.minDate));
-    const isAfterMax = this.maxDate && date.isAfter(this.parseDate(this.maxDate));
+  getCurrentMonth(month: string): boolean {
+    const date = this.activeDate.clone().month(month);
+    return date.isSame(this.current, 'month') && date.isSame(this.current, 'year') && !date.isSame(this.value, 'month');
+  }
+
+  getYearActive(year: number): boolean {
+    const date = moment(this.value).clone().year(year);
```

Comment:
Maybe don't need a clone here since the moment value is created and immediately destroyed.